File: py/tests/sdk.py
# import os


import voxgig_struct

class StructUtils:
    def __init__(self):
        # Copy all attributes from the voxgig_struct module to this class
        for attr_name in dir(voxgig_struct):
            # Skip private attributes and modules
            if not attr_name.startswith('_'):
                setattr(self, attr_name, getattr(voxgig_struct, attr_name))

class Utility:
    def __init__(self, opts=None):
        self._opts = opts
        self.struct = StructUtils()

    def check(self, ctx):
        zed = "ZED"
    
        if self._opts is None:
            zed += ""
        else:
            foo = self._opts.get("foo")
            zed += "0" if foo is None else str(foo)

        zed += "_"
        zed += str(ctx.get("bar"))

        return {"zed": zed}


class SDK:
    def __init__(self, opts=None):
        self._opts = opts or {}
        self._utility = Utility(opts)
        
    @staticmethod
    def test(opts=None):
        return SDK(opts)
        
    def tester(self, opts=None):
        return SDK(self.opts if None == opts else opts)
        
    def utility(self):
        return self._utility

File: php/tests/SDK.php
<?php

namespace Voxgig\Struct;

class SDK {
    private array $opts;
    private object $utility;

    public function __construct(array $opts = []) {
        $this->opts = $opts ?: [];
        // Capture opts for use in the closure.
        $optsCopy = $this->opts;
        $this->utility = (object)[
            // An anonymous adapter that forwards method calls to the Struct class.
            'struct' => new class {
                public function __call(string $name, array $args) {
                    // Map method name (if needed) here; otherwise, call directly.
                    return call_user_func_array(['Voxgig\Struct\Struct', $name], $args);
                }
            },
            // A simple check function similar to the TS version.
            'check' => function($ctx) use ($optsCopy) {
                $foo = isset($optsCopy['foo']) ? $optsCopy['foo'] : '';
                $bar = isset($ctx->bar) ? $ctx->bar : '0';
                return (object)[
                    'zed' => 'ZED' . $foo . '_' . $bar
                ];
            }
        ];
    }

    // Static method to obtain a test SDK instance.
    public static function test(array $opts = []): SDK {
        return new SDK($opts);
    }

    // Instance method (if needed) that mimics the async test() from TS.
    public function testMethod(array $opts = []): SDK {
        return new SDK($opts);
    }

    public function utility(): object {
        return $this->utility;
    }
}
